### PR TITLE
Add functional tests for generic:document:write event

### PR DESCRIPTION
**Pull request by Aschen**
## What does this PR do ?

:warning: Depends on https://github.com/kuzzleio/kuzzle/pull/1499

Add functional tests for generic document events.

### How should this be manually tested?

  - Step 1 : `./node_modules/.bin/cucumber features-sdk/PluginEvents.feature`

### Other changes

  - Improve functional tests speed by removing `refresh: 'wait_for'` (`42s` => `20s`)
  - test `document:delete` with the new framework

### CODE REVIEW

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1546](https://github.com/kuzzleio/kuzzle/blob/6a382715738656157b92ef1c3b24d56c226783ca/lib/services/elasticsearch.js#L1546): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1587](https://github.com/kuzzleio/kuzzle/blob/6a382715738656157b92ef1c3b24d56c226783ca/lib/services/elasticsearch.js#L1587): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1665](https://github.com/kuzzleio/kuzzle/blob/6a382715738656157b92ef1c3b24d56c226783ca/lib/services/elasticsearch.js#L1665): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1547](https://github.com/kuzzleio/kuzzle/blob/bfb146ec3e8074a21560a0dd372676a484d1f7fe/lib/services/elasticsearch.js#L1547): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1588](https://github.com/kuzzleio/kuzzle/blob/bfb146ec3e8074a21560a0dd372676a484d1f7fe/lib/services/elasticsearch.js#L1588): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1666](https://github.com/kuzzleio/kuzzle/blob/bfb146ec3e8074a21560a0dd372676a484d1f7fe/lib/services/elasticsearch.js#L1666): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: doc/2/plugins/plugin-context/constructors/elasticsearch/index.md [hunk 1]
```diff
@@ -8,9 +8,9 @@ title: Elasticsearch
 
 Instantiates an embedded [Elasticsearch client](https://github.com/elastic/elasticsearch-js) with the same configuration as the one provided in the `.kuzzlerc` file.  
 
-This client can be used to send raw Elasticsearch requests.  
+This client can be used to send raw Elasticsearch request.  
 
-See [Elasticsearch official documentation](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html) for more information.
+See [Elasticsearch official documentation](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html) for more informations.
```
Comment (Yoann-Abbes) [inline doc/2/plugins/plugin-context/constructors/elasticsearch/index.md]:
```suggestion
See [Elasticsearch official documentation](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html) for more information.
```
Never takes a `s`

**Comment by Yoann-Abbes** [comment]
I understand removing `{refresh : wait_for}` improves speed, but is it safe considering we could have some scenario like :
```
I delete the document
I check if it still exists
```
This kind of scenario could fail without this option if we check its existence with a search: as i understand it, it will check if its still indexed. What do you think?

**Comment by Aschen** [comment]
> I understand removing `{refresh : wait_for}` improves speed, but is it safe considering we could have some scenario like :
> 
> ```
> I delete the document
> I check if it still exists
> ```
> 
> This kind of scenario could fail without this option if we check its existence with a search: as i understand it, it will check if its still indexed. What do you think?

The `refresh: 'wait_for'` option make sense only when you want to search data in the collection. If you work with document IDs, there is no need to refresh the collection.  
The purpose is to always use document IDs when it's possible so we don't need to wait for the next refresh (can be 1sec) and when we need to search (count method for example), we first refresh the collection with `collection:refresh`

File: features-sdk/step_definitions/documents-steps.js [hunk 2]
```diff
@@ -44,8 +43,26 @@ Then('I {string} the following documents:', async function (action, dataTable) {
...
+  const body = this.parseObject(dataTable);
+
+  if (action === 'create') {
+    this.props.result = await this.sdk.document[action](
+      this.props.index,
+      this.props.collection,
+      body,
+      _id);
+  }
+  else {
```
Comment (alexandrebouthinon) [inline features-sdk/step_definitions/documents-steps.js]: Nit: Useless `else` if you add a `return;` in the previous block

File: features-sdk/step_definitions/documents-steps.js [hunk 3]
```diff
@@ -122,8 +139,7 @@ Then('I {string} the following document ids:', async function (action, dataTable
   this.props.result = await this.sdk.document[action](
     this.props.index,
     this.props.collection,
-    ids,
-    { refresh: 'wait_for' });
```
Comment (alexandrebouthinon) [inline features-sdk/step_definitions/documents-steps.js]: Why delete the `refresh: 'wait_for'` option?

**Pull request merged**